addons/binding/org.openhab.binding.plugwise/src/main/java/org/openhab/binding/plugwise/internal/PlugwiseCommunicationHandler.java

```diff
@@ -0,0 +1,591 @@
+        if (!isExistingMessageListener(listener, macAddress)) {
+            filteredListeners.add(new FilteredMessageListener(listener, macAddress));
+        }
+    }
+
+    /**
+     * Initialize this device and open the serial port
+     *
+     * @throws PlugwiseInitializationException if port can not be opened
+     */
+    @SuppressWarnings("rawtypes")
+    private void initializeSerialPort() throws PlugwiseInitializationException {
+
+        CommPortIdentifier portId = null;
+
+        // Parse ports and if the default port is found, initialized the reader
+        Enumeration portList = CommPortIdentifier.getPortIdentifiers();
+        while (portList.hasMoreElements()) {
+            CommPortIdentifier id = (CommPortIdentifier) portList.nextElement();
+            if (id.getPortType() == CommPortIdentifier.PORT_SERIAL
+                    && id.getName().equals(configuration.getSerialPort())) {
+                logger.debug("Serial port '{}' has been found", configuration.getSerialPort());
+                portId = id;
+                break;
+            }
+        }
+
+        if (portId == null) {
+            StringBuilder sb = new StringBuilder();
+            portList = CommPortIdentifier.getPortIdentifiers();
+            while (portList.hasMoreElements()) {
+                CommPortIdentifier id = (CommPortIdentifier) portList.nextElement();
+                if (id.getPortType() == CommPortIdentifier.PORT_SERIAL) {
+                    sb.append(String.format("%s%n", id.getName()));
+                }
+            }
+            throw new PlugwiseInitializationException(
+                    String.format("Serial port '%s' could not be found. Available ports are:%n%s",
+                            configuration.getSerialPort(), sb));
+        }
+
+        try {
+            serialPort = portId.open("openHAB", 2000);
+            serialPort.addEventListener(this);
+            serialPort.notifyOnDataAvailable(true);
+            serialPort.setSerialPortParams(115200, SerialPort.DATABITS_8, SerialPort.STOPBITS_1,
+                    SerialPort.PARITY_NONE);
+            outputChannel = Channels.newChannel(serialPort.getOutputStream());
+        } catch (IOException e) {
+            throw new PlugwiseInitializationException("Failed to get serial port output stream", e);
+        } catch (PortInUseException e) {
+            throw new PlugwiseInitializationException("Serial port already in use", e);
+        } catch (TooManyListenersException e) {
+            throw new PlugwiseInitializationException("Failed to add serial port listener", e);
+        } catch (UnsupportedCommOperationException e) {
+            throw new PlugwiseInitializationException("Failed to set serial port parameters", e);
+        }
+    }
+
+    public boolean isExistingMessageListener(PlugwiseMessageListener listener) {
+        for (FilteredMessageListener filteredListener : filteredListeners) {
```

Comment:
It has been reworked to use lambdas. The code now lives in the PlugwiseFilteredMessageListenerList.